File: ra-idp/src/authentication/ports/mfa-factor-repository.ts
/**
 * Layer 3 — Application Logic（ポート定義）
 *
 * spec/scl.yaml MfaFactor 集合へのアクセス。identity は (sub, type) 複合キーで、
 * 1 ユーザに各 type 高々 1 個まで持てる。User.mfa_enrolled は本集合の非空性から
 * 導出する（sub に何らかの factor があれば true）。
 */

import type { MfaFactor, MfaFactorType } from '../../spec-bindings/schemas'

export interface MfaFactorRepository {
  /** 指定ユーザの全 factor を返す。順序は実装依存。 */
  listBySub(sub: string): Promise<MfaFactor[]>
  /** 指定 (sub, type) の factor を返す。存在しなければ null。 */
  find(sub: string, type: MfaFactorType): Promise<MfaFactor | null>
  /** factor を保存する。同じ (sub, type) があれば置き換える (upsert)。 */
  save(factor: MfaFactor): Promise<void>
  /** 指定 (sub, type) の factor を削除する。存在しなくてもエラーにしない。 */
  delete(sub: string, type: MfaFactorType): Promise<void>
}

File: ra-idp-go/internal/authentication/ports/mfa_factor_repository.go
package ports

import (
	"context"

	"ra-idp-go/internal/shared/spec"
)

type MfaFactorRepository interface {
	ListBySub(ctx context.Context, sub string) ([]*spec.MfaFactor, error)
	Find(ctx context.Context, sub string, factorType spec.MfaFactorType) (*spec.MfaFactor, error)
	Save(ctx context.Context, factor *spec.MfaFactor) error
	Delete(ctx context.Context, sub string, factorType spec.MfaFactorType) error
	// DeleteAllForSub は ADR-036 の anonymize cascade から呼ばれる。
	// 対象 sub の MFA factor をすべて物理削除する。
	DeleteAllForSub(ctx context.Context, sub string) error
}
